releasenotes/notes/vf2-post-layout-f0213e2c7ebb645c.yaml

```diff
@@ -0,0 +1,47 @@
+---
+features:
+  - |
+    Added a new transpiler pass, :class:`~.VF2PostLayout`. This pass is of a
+    new type to perform a new phase/function in the compilation pipeline,
+    post-layout or post optimization qubit selection. The idea behind this
+    pass is after we finish the optimization loop in transpiler we have a
+    know what the final gate counts will be on each qubit in the circuit so
+    we can potentially find a better performing subset of qubits on a backend
+    to execute the circuit. The pass will search for an isomorphic subgraph in
+    the connectivity graph of the target backend and look at the full error
+    rate of the complete circuit on any subgraph found and use return the
+    layout found with the lowest error rate for the circuit.
+
+    This pass is similar to the :class:`~.VF2Layout` pass and both internally
+    use the same VF2 implementation from 
+    `retworkx <https://github.com/Qiskit/retworkx>`__. However,
+    :class:`~.VF2PostLayout` is deisgned to run after initial layout, routing,
+    basis translation, and any optimization passes run and will only work if
+    a layout has already been applied, the circuit has been routed, and all
+    gates are in the target basis. This is required so that when a new layout
+    is applied the circuit can still be run on the target device. :class:`~.VF2Layout`
+    on the other hand is designed to find a perfect initial layout and can
+    work with any circuit.
+  - |
+    The :class:`~.ApplyLayout` transpiler pass now has support for updating
+    a layout on a circuit after a layout has been applied once before. If
+    the ``post_layout`` field is present (in addition to the required
+    ``layout`` field) the ``property_set`` when the :class:`~.ApplyLayout` pass
+    is run the pass will update the layout to apply the new layout. This will
+    return a :class:`~.DAGCircuit` with the qubits in the new physical order
+    and the ``layout`` property set will be updated so that it maps the
+    virtual qubits from the original layout to the physical qubits in the new
+    ``post_layout`` field.
```

Comment:
We need both so that we can update the layout from the original virtual bits in the input circuit to the new physical bits. The problem is that by the point in the pipeline we run `VF2PostLayout` the virtual qubits from the input circuit have been removed and we've replaced the circuit with a flat register with the num_qubits from the backend. This causes a problem when we go to apply the layout after, if we re-use layout we can either set the virtual bits in the layout to the bits from the circuit and apply layout will work but then we lose the annotation of the original input circuit virtual bit mapping. Or we can use the original virtual bits in the new layout which will preserve that information but then apply layout will not do anything because none of the old bits are present in the dag at this point. By having both in apply layout we can reorder the bits in the circuit and then also update the output layout to map from old virtual -> new physical.